83 implement stage 5 in dashboard

azywot opened the pull request:
Added stage 5 to the dashboard:
- `% ACTIVITION`/`% INHIBITION`/`Z-SCORE` visualizations 
- interactive Z-score adjustment
- data table with compounds and their key columns

**NOTE: I think we should add one more stage for generating raports etc. (updated last stages visible [here](https://docs.google.com/presentation/d/115voct6UoRCp64WlLkQB6gekUEL3kYhD/edit#slide=id.p12))**

This is what the page looks like for now, the area plot we mentioned during the meeting is TBD
![obraz](https://github.com/zuzg/drug-screening/assets/82370491/42587903-07d1-4f14-9e3f-21ce46f4c205)

Here I used all the data we have from Primary Screening within one assay as an example:
![obraz](https://github.com/zuzg/drug-screening/assets/82370491/4f378605-d546-44f0-a610-ea4e778bd74e)
For the future version of the plot interactivity will be retained only for compounds below/above the thresholds + stats for the area plot will be added (as it gets laggy and interactivity becomes unnecessary)

**NOTE 2: For the plots I tried different layouts on a grid with the data table but ultimately decided it is the most readable like this. However, I'm not an expert when it comes to styling ngl, therefore any suggestions appreciated 🙏🙏**

Tremirre commented on dashboard/data/combine.py:
```diff
@@ -108,7 +110,7 @@ def combine_bmg_echo_data(
     df_stats: pd.DataFrame,
     plate_values: np.ndarray,
-    modes: dict[Mode],
+    modes: dict[Mode] = dict(),
```
mutable objects shouldn't be set as defaults in functions as this introduces a subtle bug in form of them being shared among all function calls in the program

instead if you want to set an empty dict as default you should write:

```python
def func(
    ...
    should_default_to_dict: dict | None = None,
    ...
) -> ...:
    if should_default_to_dict is None:
        should_default_to_dict = {}
    ...
```

this way each function call creates its own dictionary

Tremirre commented on dashboard/pages/primary_screening/stages/s5_summary.py:
```diff
@@ -1,4 +1,47 @@
+        "width": "70px",
+        "autosize": {"type": "fit", "resize": True},
+        "overflow": "hidden",
+    },
```
suggestion: consider adding `style_cell` argument as:
```python
style_cell={
    "font-family": "sans-serif",
    "font-size": "12px",
}
```
to use nicer font (imo more readable - I use it in my stage)

Tremirre commented on dashboard/pages/primary_screening/callbacks.py:
```diff
@@ -217,6 +223,102 @@ def upload_echo_data(contents, names, last_modified, stored_uuid, file_storage):
 # === STAGE 5 ===
 
 
+def on_stage_5_entry(current_stage: int, stored_uuid: str, file_storage: FileStorage):
```
suggestion: use explicit stage name instead of index in the callback name - it makes it more clear what should be in the callback and in general what it refers to, and the index itself may be inferred from the `if current_stage` check

azywot replied:
Good point, will do that

Tremirre reviewed (commented):
Nice job, implementation seems good but I have a couple of suggestions regarding the functionality and general observations:

1. Z-score range

For this task I'd suggest using [two-sided slider](https://dash.plotly.com/dash-core-components/rangeslider) for better UX, and since the plots are quite computationally heavy (meaning each individual change forces a long reaload) you could consider adding a `recalculate` or `apply` button that would read the slider change and then regenerate the graphs

2. Plots

All three plots are quite data-heavy and take a lot of space - they all do not fit on a single page - hence I'd recommend just showing one plot at a time with a SelectionBox that would select which visualization to show (similarily as in DrugScreeningDashboard v1.0.0 with projections)

Then you'll be able to continue our struggle to make the stage not overflow the page which is nicer for the user

3. Dash DataTable

For consistency I'd recommend you copy the styling from stage 2 datatable so that they look alike (alternating rows with different colors, different font and text size) 
Furthermore we might consider if showing all 100k+ compounds in a table is really necessary - maybe we should restrict that only to the special ones

4. Questions
- Why sometimes Activation % and Inhibition % values are different? (I do not remember if that was the case or not previously)
- Why is there Z-filtering being done twice? (Once in stage 3 by @zuzg and once here)

azywot commented:
> Nice job, implementation seems good but I have a couple of suggestions regarding the functionality and general observations:
> ## 1. Z-score range
> 
> For this task I'd suggest using [two-sided slider](https://dash.plotly.com/dash-core-components/rangeslider) for better UX, and since the plots are quite computationally heavy (meaning each individual change forces a long reaload) you could consider adding a `recalculate` or `apply` button that would read the slider change and then regenerate the graphs
> ## 2. Plots
> 
> All three plots are quite data-heavy and take a lot of space - they all do not fit on a single page - hence I'd recommend just showing one plot at a time with a SelectionBox that would select which visualization to show (similarily as in DrugScreeningDashboard v1.0.0 with projections)
> 
> Then you'll be able to continue our struggle to make the stage not overflow the page which is nicer for the user
> ## 3. Dash DataTable
> 
> For consistency I'd recommend you copy the styling from stage 2 datatable so that they look alike (alternating rows with different colors, different font and text size) Furthermore we might consider if showing all 100k+ compounds in a table is really necessary - maybe we should restrict that only to the special ones
> ## 4. Questions
> 
>     * Why sometimes Activation % and Inhibition % values are different? (I do not remember if that was the case or not previously)
> 
>     * Why is there Z-filtering being done twice? (Once in stage 3 by @zuzg and once here)

Thanks for the comments/suggestions 😊
Ad. 1, 2 and 3 - I've added a `RangeSlider` as you suggested, and in the meantime put the plots to separate tabs. The same styling as in the case of a data table from `stage 2` was also added. Below I'm pasting the preview:
![obraz](https://github.com/zuzg/drug-screening/assets/82370491/07fa4cfc-d837-41c6-b847-a7df16380bc5)
Ad. 4
- I was actually surprised that they're mosty of the same value, I've already double checked with the pre-calculated spreadsheets we received but found only overlapping values in `% ACTIVATION` case (no data to compare for `% INHIBITION`). Great that you spotted this one, we'll have to ask about some additional data to compare and decide upon handling such cases as *calculate only activation/inhibition for this assay, use this formula for activation (there are two)... etc*. Have these in the back of my mind, for the future need to be taken care of.
- this is `Z-score` here whereas @zuzg took care of `Z/Z' factors`, according to [Dashboard for data analysis](https://docs.google.com/presentation/d/1OF7pJa0Hh8XVtQUSG58z_Jo2WRNNZOf9/edit#slide=id.p5)

zuzg commented on dashboard/pages/primary_screening/callbacks.py:
```diff
@@ -1,19 +1,25 @@
+    visualize_multiple_plates,
+)
 
+from ...data.bmg_plate import parse_bmg_files
+from ...data.file_preprocessing.echo_files_parser import EchoFilesParser
```
duplicated imports (just remove these lines)

zuzg commented on dashboard/visualization/plots.py:
```diff
@@ -268,101 +268,100 @@ def visualize_activation_inhibition_zscore(
-    :param control_pos_df: DataFrame with positive control values
-    :param control_neg_df: DataFrame with negative control values
-    :param column: values to visualise
+    Visualize activation/inhibition z-score for each compound
```
missing newline before params

zuzg reviewed (commented):
Nice job! The page looks very good and separating plots into tabs definitely improved readability. Please make sure that after merging develop (there are some conflicts) all the pages work correctly  ☺️

azywot commented:
> Nice job! The page looks very good and separating plots into tabs definitely improved readability. Please make sure that after merging develop (there are some conflicts) all the pages work correctly ☺️

Thanks, I've just merged and went through all the stages, everything works as expected 😊

Tremirre reviewed (approved):
LGTM, thanks for the reply!
We're good to merge
